### PR TITLE
added tri / tril / triu

### PR DESCRIPTION
**Pull request by dc-dc-dc**
## Proposed changes

Adding tri / tril / triu

I know there was pushback on using torch as a dependency in tests but was wondering if comparing to numpy was okay, simplifies and expands the test cases to try against vs saving expected results.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
> but was wondering if comparing to numpy was okay

Yup that's ok, we already do it in fact.

**Comment by dc-dc-dc** [comment]
Cool, added some basic ctests and utilized numpy in the python tests which cover more diagonals.

File: mlx/ops.h [hunk 1]
```diff
@@ -110,6 +110,10 @@ inline array identity(int n, StreamOrDevice s = {}) {
   return identity(n, float32, s);
 }
 
+array tri(int n, int m, int k, Dtype type, StreamOrDevice s = {});
```
Comment (awni) [inline mlx/ops.h:113]: Can you do an overload of  `tri(int n, Dtype ...) where `m=n` and `k=0`. I think it's a common use case.

File: python/tests/test_ops.py [hunk 2]
```diff
@@ -320,6 +320,22 @@ def test_array_equal(self):
                 self.assertFalse(mx.array_equal(x, y))
                 self.assertTrue(mx.array_equal(x, y, equal_nan=True))
 
+    def test_tri(self):
+        for diag in [-1, 0, 1, -2]:
+            self.assertEqualArray(mx.tri(4, 4, diag), mx.array(np.tri(4, 4, diag)))
+
+    def test_tril(self):
+        mt = mx.random.normal((10, 10))
+        nt = np.array(mt.tolist())
```
Comment (awni) [inline python/tests/test_ops.py:329]: `.tolist` is not necesssary

File: python/tests/test_ops.py [hunk 3]
```diff
@@ -320,6 +320,22 @@ def test_array_equal(self):
...
+        for diag in [-1, 0, 1, -2]:
+            self.assertEqualArray(mx.tril(mt, diag), mx.array(np.tril(nt, diag)))
+
+    def test_triu(self):
+        mt = mx.random.normal((10, 10))
+        nt = np.array(mt.tolist())
```
Comment (awni) [inline python/tests/test_ops.py]: same here.

File: python/tests/test_ops.py [hunk 4]
```diff
@@ -320,6 +320,22 @@ def test_array_equal(self):
                 self.assertFalse(mx.array_equal(x, y))
                 self.assertTrue(mx.array_equal(x, y, equal_nan=True))
 
+    def test_tri(self):
+        for diag in [-1, 0, 1, -2]:
+            self.assertEqualArray(mx.tri(4, 4, diag), mx.array(np.tri(4, 4, diag)))
+
```
Comment (awni) [inline python/tests/test_ops.py:326]: please add one or two test for `m` and `k` other than defaults.

File: python/src/ops.cpp [hunk 5]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
+      "tril",
+      &tril,
+      "x"_a,
+      "k"_a = 0,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+      tril(x: array, k: int, *, stream: Union[None, Stream, Device] = None) -> array
+        
+        An array with zeros above the given diagonal and self elsewhere.
```
Comment (awni) [inline python/src/ops.cpp:1449]: `self elsewhere` is odd. Maybe change this to `Zeros the array above the given diagonal.`

File: python/src/ops.cpp [hunk 6]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
+
+        Args:
+          x (array): input array.
+          k (int): The diagonal of the 2-D array
+          stream (Stream, optional): Stream or device. Defaults to None.
```
Comment (awni) [inline python/src/ops.cpp:1454]: Double back ticks on `None`

File: python/src/ops.cpp [hunk 7]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
+        
+        Returns:
+          array: Array filled with zeros above given diagonal
```
Comment (awni) [inline python/src/ops.cpp:1457]: `Array zeroed above the given diagonal`

File: python/src/ops.cpp [hunk 8]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
+        
+        An array with zeros below the given diagonal and self elsewhere.
+
+        Args:
+          x (array): input array.
+          k (int): The diagonal of the 2-D array
+          stream (Stream, optional): Stream or device. Defaults to None.
+        
+        Returns:
+          array: Array filled with zeros below given diagonal
```
Comment (awni) [inline python/src/ops.cpp:1477]: Same comments for this docstring but below instead of above.

File: python/src/ops.cpp [hunk 9]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
           array: An identity matrix of size n x n.
       )pbdoc");
+  m.def(
+      "tri",
+      [](int n, int m, int k, std::optional<Dtype> dtype, StreamOrDevice s) {
+        return tri(n, m, k, dtype.value_or(float32), s);
+      },
+      "n"_a,
+      "m"_a,
+      "k"_a,
```
Comment (awni) [inline python/src/ops.cpp:1420]: Could you make these optional and follow the numpy defaults?

File: python/src/ops.cpp [hunk 10]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
+        tri(n: int, m: int, k: int, dtype: Optional[Dtype] = None, *, stream: Union[None, Stream, Device] = None) -> array
+        
+        An array with ones at and below the given diagonal and zeros elsewhere.
+
+        Args:
+          n (int): The number of rows in the output.
+          m (int): The number of cols in the output.
+          k (int): The diagonal of the 2-D array
+          dtype (Dtype, optional): Data type of the output array. Defaults to float32.
+          stream (Stream, optional): Stream or device. Defaults to None.
```
Comment (awni) [inline python/src/ops.cpp:1434]: Double backticks on `float32` and `None`.

File: python/src/ops.cpp [hunk 11]
```diff
@@ -1410,6 +1410,72 @@ void init_ops(py::module_& m) {
...
       )pbdoc");
+  m.def(
+      "tri",
+      [](int n, int m, int k, std::optional<Dtype> dtype, StreamOrDevice s) {
+        return tri(n, m, k, dtype.value_or(float32), s);
+      },
+      "n"_a,
+      "m"_a,
+      "k"_a,
+      "dtype"_a = std::nullopt,
```
Comment (awni) [inline python/src/ops.cpp:1421]: This could just default to `float32` rather than use an `optional` type. Simpler that way.

Comment (awni) [reply python/src/ops.cpp:1421]: I should have caught that for some other ops like `eye` above. If you are up for cleaning those as well that would be nice!

File: mlx/ops.cpp [hunk 12]
```diff
@@ -218,6 +218,28 @@ array identity(int n, Dtype dtype, StreamOrDevice s /* = {} */) {
...
+  if (x.ndim() < 2) {
+    throw std::invalid_argument("array must be atleast 2-D");
+  }
+  auto mask = tri(x.shape(-2), x.shape(-1), k, x.dtype(), s);
+  return where(mask, x, zeros_like(x, s), s);
+}
+
+array triu(array x, int k, StreamOrDevice s /* = {} */) {
+  if (x.ndim() < 2) {
+    throw std::invalid_argument("array must be atleast 2-D");
```
Comment (awni) [inline mlx/ops.cpp:237]: Add `[triu]` to the front of the error message.

File: mlx/ops.cpp [hunk 13]
```diff
@@ -218,6 +218,28 @@ array identity(int n, Dtype dtype, StreamOrDevice s /* = {} */) {
...
 
+array tri(int n, int m, int k, Dtype type, StreamOrDevice s /* = {} */) {
+  auto l = broadcast_to(expand_dims(arange(n, s), 1, s), {n, m}, s);
+  auto r = broadcast_to(expand_dims(arange(-k, m - k, s), 0, s), {n, m}, s);
+  return astype(greater_equal(l, r, s), type, s);
+}
+
+array tril(array x, int k, StreamOrDevice s /* = {} */) {
+  if (x.ndim() < 2) {
+    throw std::invalid_argument("array must be atleast 2-D");
```
Comment (awni) [inline mlx/ops.cpp:229]: add `[tril]` to the front of the error message.

File: mlx/ops.cpp [hunk 14]
```diff
@@ -218,6 +218,28 @@ array identity(int n, Dtype dtype, StreamOrDevice s /* = {} */) {
   return eye(n, n, 0, dtype, s);
 }
 
+array tri(int n, int m, int k, Dtype type, StreamOrDevice s /* = {} */) {
+  auto l = broadcast_to(expand_dims(arange(n, s), 1, s), {n, m}, s);
+  auto r = broadcast_to(expand_dims(arange(-k, m - k, s), 0, s), {n, m}, s);
```
Comment (awni) [inline mlx/ops.cpp:223]: explicit broadcasting is not needed here, `greater_equal` broadcasts its args.

**Review by awni — CHANGES_REQUESTED** [review]
This is really awesome, thanks for adding it! Left some comments.

**Comment by nullhook** [comment]
waiting for this to be merged

**Comment by awni** [comment]
Just added a few minor updates + docs. This is ready to go. Will merge as soon as the CI is done!

**Review by awni — APPROVED** [review]
Beautifully done, thank you!

**Pull request merged**